ts/src/connections/tunnelRelayTunnelClient.ts

```diff
@@ -1,87 +1,73 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
-import {
-    TunnelConnectionMode,
-    Tunnel,
-    TunnelEndpoint,
-    TunnelRelayTunnelEndpoint,
-    TunnelAccessScopes,
-} from '@vs/tunnels-contracts';
-import { TunnelAccessTokenProperties } from '@vs/tunnels-management';
-import { TraceLevel } from '@vs/vs-ssh';
-import { TunnelRelayStreamFactory, DefaultTunnelRelayStreamFactory } from '.';
+import { TunnelConnectionMode, Tunnel, TunnelRelayTunnelEndpoint } from '@vs/tunnels-contracts';
+import { CancellationToken } from 'vscode-jsonrpc';
+import { TunnelManagementClient } from '@vs/tunnels-management';
+import { Stream, Trace } from '@vs/vs-ssh';
 import { TunnelClientBase } from './tunnelClientBase';
+import { tunnelRelaySessionClass } from './tunnelRelaySessionClass';
+
+const webSocketSubProtocol = 'tunnel-relay-client';
 
 /**
  * Tunnel client implementation that connects via a tunnel relay.
  */
-export class TunnelRelayTunnelClient extends TunnelClientBase {
-    /**
-     * Web socket sub-protocol to connect to the tunnel relay endpoint.
-     */
-    public static webSocketSubProtocol = 'tunnel-relay-client';
+export class TunnelRelayTunnelClient extends tunnelRelaySessionClass(
+    TunnelClientBase,
+    webSocketSubProtocol,
+) {
     public connectionModes: TunnelConnectionMode[] = [];
 
-    /**
-     * Gets or sets a factory for creating relay streams.
-     */
-    public streamFactory: TunnelRelayStreamFactory = new DefaultTunnelRelayStreamFactory();
-
-    constructor() {
-        super();
+    constructor(trace?: Trace, managementClient?: TunnelManagementClient) {
+        super(trace, managementClient);
     }
 
-    public async connectClient(tunnel: Tunnel, endpoints: TunnelEndpoint[]): Promise<void> {
-        let tunnelEndpoints = endpoints.map((endpoint) => endpoint as TunnelRelayTunnelEndpoint);
+    /**
+     * Gets the tunnel relay URI.
+     */
+    public async getTunnelRelayUri(tunnel?: Tunnel): Promise<string> {
+        let tunnelEndpoints = this.endpoints!.map(
+            (endpoint) => endpoint as TunnelRelayTunnelEndpoint,
```

Comment:
This `map()` call will return `null` items for enpoints that are not the expected type. But it is intended to filter them out instead. So, this should be changed to `this.endpoints!.filter((endpoint) => endpoint instanceof TunnelRelayTunnelEndpoint)`

(I see that this bug was already there before the PR.)

Reply:
`instanceof` cannot be used because `TunnelRelayTunnelEndpoint` is an interface. Instead, we can use `endpoint.connectionMode === TunnelConnectionMode.TunnelRelay` condition.